pkg/lint/lintersdb/manager.go

```diff
@@ -414,10 +414,17 @@ func (m Manager) GetAllSupportedLinterConfigs() []*linter.Config {
 			WithLoadForGoAnalysis().
 			WithURL("https://github.com/kyoh86/exportloopref"),
 
-		linter.NewConfig(golinters.NewForbidigo(forbidigoCfg)).
-			WithSince("v1.34.0").
-			WithPresets(linter.PresetStyle).
-			WithURL("https://github.com/ashanbrown/forbidigo"),
+		func() *linter.Config {
+			l := linter.NewConfig(golinters.NewForbidigo(forbidigoCfg)).
+				WithSince("v1.34.0").
+				WithPresets(linter.PresetStyle).
+				WithURL("https://github.com/ashanbrown/forbidigo")
+			if forbidigoCfg != nil && forbidigoCfg.AnalyzeTypes {
+				// Need more information for the analysis.
+				l = l.WithLoadForGoAnalysis()
```

Comment:
@ldez As @pohly mentioned, we are making this call (which sets `IsSlow`) conditional on the configuration.  Are you ok with having this otherwise "fast" linter become slow if this flag is set or would that be considered too unexpected?

Reply:
`GetAllSupportedLinterConfigs` is called in different situations, the linter config will not be available in all those situations, so please remove this and don't try to create an asymetrical configuration between `WithLoadForGoAnalysis` and `LoadModeTypesInfo`.

Reply:
I see how not setting `IsSlow` in `GetAllSupportedLinterConfigs` when called without config is wrong (or at least potentially misleading).

But is it also useful in `NewForbidigo` to use `LoadModeTypesInfo` whether it's needed or not? Can something break when combining `WithLoadForGoAnalysis` with `LoadModeSyntax`?

Reply:
Okay. I've pushed an update which always uses the more expensive mode. In the long run I'd expect users to switch to it anyway. It's just not the default because of some risk of breaking certain patterns.

Reply:
For the record, I'm not sure I agree, and I have to admit I'm now starting to regret my decision to allow type-based checking.  I think forbidigo is most useful if it can be run as a pre-commit check, and my guess is that this outweighs the value of being able to check these additional kinds of patterns (which still behave unexpectedly in a few cases).  I would strongly favor finding some way to make this linter behave slowly only if the `analyzeTypes` config is set.

Reply:
Is anyone really running only "fast" linters in a pre-commit check and the "slow" ones after merging? That seems odd because then mistakes get merged and need to be fixed after the fact.

I can see the advantage of running only "fast" linters while developing the code to speed up the write/compile/test loop, but I am not sure how important it is to include forbidigo in that loop.

Reply:
I don't think the pattern of running a fast set of checks on pre-commit is too unusual.  It's one I've employed frequently.  As the most extreme example, you could likely run all your tests on pre-commit, but we tend to do that only in CI because they're too slow.  The example cases of `fmt.Println` and `ginkgo.FDescribe` are the kinds of thing that are great to catch on pre-commit.  It's nice not to have to wait for a whole CI cycle to know that you left in some debug code.

Reply:
pre-commit is not something so usual because pre-commit has several problems (cannot be committed, require an installation of the hook, etc.).
Only running "fast" linters and using the CI to run "slow" linters is also not so usual.

Personally and professionally, I don't use pre-commit, and I don't want to wait for a CI to get feedback from the linter (CIs are slower than my computer).

Reply:
Now I think I understand what @ashanbrown meant with "pre-commit". I don't force linter runs when committing code either. Locally, I run linters when I am ready to submit something (unless I forget...) or when I believe that it might find something.

In Kubernetes, we run all linters when testing a pull request, and those results are merge-blocking, i.e. we never merge code that has linter issues. That's what I meant earlier.

Reply:
Back to the main point of this discussion: @ldez, you had asked that we not be asymmetric in our use of `WithLoadForGoAnalysis` in one place and not the other.  Could we simply report that we need the requirements of a fast linter except when this "experimental" setting is set in the config file?  When we see the experimental config, we'd add the extra load requirements.  I realize this is still asymmetric, although at least it would no longer call `WithLoadForGoAnalysis`.  Is the concern that by reporting a linter as fast, say in a help message, and later making it slow, we would be confusing the user, or is there something that would actually break?

Reply:
`go test -v -run TestSourcesFromTestdata/forbidigo ./test` doesn't pass when `GetAllSupportedLinterConfigs` doesn't use `WithLoadForGoAnalysis` for forbidigo, so it has technical implications.

If we absolutely need a "fast forbidigo" (not convinced yet myself, but usage patterns are different and I don't want to make forbidigo less useful for its original author!) and a "throrough forbidigo" (a strong yes from me for Kubernetes, because import aliases make the current source-code based rules unreliable), then we could register forbidigo twice:
- Once with the current configuration and mode of operation (no changes).
- Again with a different name (working title "forbidigo-analyze-types", better suggestions welcome) and the new configuration without the `AnalyzeTypes` field.

Reply:
I'm not sure it's `LoadModeTypesInfo` that is causing us to be considered a slow linter.  For example, durationCheck uses this load mode and is still considered fast.  I think it's the requirement to load dependencies that would make this a slow linter.  `WithLoadForGoAnalysis` does this:

```
lc.LoadMode |= packages.NeedImports | packages.NeedDeps | packages.NeedExportFile | packages.NeedTypesSizes
```

This was added at https://github.com/golangci/golangci-lint/pull/1844.  Perhaps @ldez can fill us on as to why we can't just set `IsSlow` based on the presence of select LoadMode flags.

I'd personally be happy with the multiple-linter configs but that had sounded like a non-starter to me.

Reply:
`goanalysis.LoadModeTypesInfo` required `WithLoadForGoAnalysis()`.

A linter with two "speed states" doesn't seem right.

"slow" or "fast" don't really describe the speed of a linter, it's just a best-effort state.
The "slow" linters share a state for the slow part, so if you run only one linter you will pay the cost but if your run several linters the cost will be shared.

Duplicating the linter doesn't seem like the right idea too.

I think it's not really important if your linter is flagged as "slow" or "fast".
But if it's important to you, you have to make choice inside the linter: use types or not.

I don't want to create unneeded complexity inside the configuration, I also prefer maintainability over possibilities.

Reply:
Would it work to register it with `WithLoadForGoAnalysis` unconditionally and then decide in `NewForbidigo` based on the actual setting? At least the tests are passing that way.

Would it make the linter run faster? I'm not so sure.

For Kubernetes, `golangci-lint run --disable-all --enable forbidigo -v` with that approach uses `LoadModeSyntax`:
```
INFO Execution took 8.982228038s
```

That's after cleaning the golangci-lint cache, but with a warm Go build cache.

Without this hack it's 9.556500684s - a bit slower, but not much.

With an empty Go cache (`go clean --cache`), the difference is 56.58711488s (with that approach) vs. 57.399810283s (without).

For comparison, with forbidigo master (985efcc1) it's 9.639382845s and 57.23967985s.

These numbers are so close together, I don't see a relevant difference. It probably comes down to noise - otherwise master would have to be the fastest.

@ashanbrown: can you give a specific example where you want to use "fast" analysis? Does this PR really make a significant difference?